src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1151,11 +1439,47 @@ RebalanceTableShards(RebalanceOptions *options, Oid shardReplicationModeOid)
 }
 
 
+/*
+ * ConflictShardPlacementUpdateOnlyWithIsolationTesting is only useful for
+ * testing and should not be called by any code-path except for
+ * UpdateShardPlacement().
+ *
+ * To be able to test the rebalance monitor functionality correctly, we need to
+ * be able to pause the rebalancer at a specific place in time. We cannot do
+ * this by block the shard move itself someway (e.g. by calling truncate on the
+ * distributed table). The reason for this is that we do the shard move in a
+ * newly opened connection. This causes our isolation tester block detection to
+ * not realise that the rebalance_table_shards call is blocked.
+ *
+ * So instead, before opening a connection we lock an advisory lock that's
+ * based on the shard id. By locking this advisory lock in a different session
+ * we can block the rebalancer in a way that the isolation tester block
+ * detection is able to detect.
+ */
+static void
+ConflictShardPlacementUpdateOnlyWithIsolationTesting(uint64 shardId)
```

Comment:
I think it would be cleaner to adjust citus_isolation_test_session_is_blocked, but to be fair that's probably a lot more complicated.

Reply:
I did try that initially, by adding the pid of the process that opens the connection in the `application_name` (like we do for the publication/subscription stuff). But since we use the regular connection opening code, this would require lots of changes (afaict). So I decided to go with the slightly more hacky, but easier to implement approach.